File: ASP.NET/365ThreeSixty_v3/365ThreeSixty_v3/fuzzyMatch.vb
Module fuzzyMatch
    Function FuzzyPercent(ByVal String1 As String,
                      ByVal String2 As String,
                      Optional Algorithm As Integer = 3,
                      Optional Normalised As Boolean = False) As Single
        '*************************************
        '** Return a % match on two strings **
        '*************************************
        Dim intLen1 As Integer, intLen2 As Integer

        Dim intScore As Integer
        Dim intTotScore As Integer


        '-------------------------------------------------------
        '-- If strings havent been normalised, normalise them --
        '-------------------------------------------------------
        If Normalised = False Then
            String1 = LCase$(Trim(String1))
            String2 = LCase$(Trim(String2))
        End If

        '----------------------------------------------
        '-- Give 100% match if strings exactly equal --
        '----------------------------------------------
        If String1 = String2 Then
            Return 1
            Exit Function
        End If

        intLen1 = Len(String1)
        intLen2 = Len(String2)

        '----------------------------------------
        '-- Give 0% match if string length < 2 --
        '----------------------------------------
        If intLen1 < 2 Then
            Return 0
            Exit Function
        End If

        intTotScore = 0                   'initialise total possible score
        intScore = 0                      'initialise current score

        '--------------------------------------------------------
        '-- If Algorithm = 1 or 3, Search for single characters --
        '--------------------------------------------------------
        If (Algorithm And 1) <> 0 Then
            FuzzyAlg1(String1, String2, intScore, intTotScore)
            If intLen1 < intLen2 Then FuzzyAlg1(String2, String1, intScore, intTotScore)
        End If

        '-----------------------------------------------------------
        '-- If Algorithm = 2 or 3, Search for pairs, triplets etc. --
        '-----------------------------------------------------------
        If (Algorithm And 2) <> 0 Then
            FuzzyAlg2(String1, String2, intScore, intTotScore)
            If intLen1 < intLen2 Then FuzzyAlg2(String2, String1, intScore, intTotScore)
        End If

        FuzzyPercent = intScore / intTotScore

    End Function
    Private Sub FuzzyAlg1(ByVal String1 As String,
                          ByVal String2 As String,
                          ByRef Score As Integer,
                          ByRef TotScore As Integer)
        Dim intLen1 As Integer, intPos As Integer, intPtr As Integer, intStartPos As Integer

        intLen1 = Len(String1)
        TotScore = TotScore + intLen1              'update total possible score
        intPos = 0
        For intPtr = 1 To intLen1
            intStartPos = intPos + 1
            intPos = InStr(intStartPos, String2, Mid$(String1, intPtr, 1))
            If intPos > 0 Then
                If intPos > intStartPos + 3 Then     'No match if char is > 3 bytes away
                    intPos = intStartPos
                Else
                    Score = Score + 1          'Update current score
                End If
            Else
                intPos = intStartPos
            End If
        Next intPtr
    End Sub
    Private Sub FuzzyAlg2(ByVal String1 As String,
                            ByVal String2 As String,
                            ByRef Score As Integer,
                            ByRef TotScore As Integer)
        Dim intCurLen As Integer, intLen1 As Integer, intTo As Integer, intPtr As Integer, intPos As Integer
        Dim strWork As String

        intLen1 = Len(String1)
        For intCurLen = 2 To intLen1
            strWork = String2                          'Get a copy of String2
            intTo = intLen1 - intCurLen + 1
            TotScore = TotScore + Int(intLen1 / intCurLen)  'Update total possible score
            For intPtr = 1 To intTo Step intCurLen
                intPos = InStr(strWork, Mid$(String1, intPtr, intCurLen))
                If intPos > 0 Then
                    Mid(strWork, intPos, intCurLen) = intCurLen.ToString 'corrupt found string
                    Score = Score + 1     'Update current score
                End If
            Next intPtr
        Next intCurLen

    End Sub
End Module

File: 365ThreeSixtyAPI/365ThreeSixtyAPI/Factories/fuzzyMatch.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Collections;
using System.Data;
using System.Diagnostics;
using Microsoft.VisualBasic;

namespace _365ThreeSixtyAPI.Factories
{


    public class fuzzyMatch
    {
        public double FuzzyPercent(string String1, string String2, int Algorithm = 3, bool Normalised = false)
        {
            //normalise strings
            if(Normalised == false)
            {
                String1 = String1.ToLower();
                String2 = String2.ToLower();             
            }

            //give 100% if strings match exactly
            if (String1 == String2)
            {
                return 1;
            }

            //set parameters
            int intLen1 = String1.Length;
            int intLen2 = String2.Length;

            //give 0% if string length <2
            if (intLen1 < 2 || intLen2<2)
            {
                return 0;
            }

            //initilaise totalal possible score
            int intTotScore = 0;
            //initilise current score
            int intScore = 0;

            //if algorythm = 1 or 3 then search for single characters
            if ((Algorithm & 1) != 0)
            {
                if (intLen1 > intLen2)
                {
                    FuzzyAlg1(String1, String2, ref intScore, ref intTotScore);
                }
                else
                {
                    FuzzyAlg1(String2, String1, ref intScore, ref intTotScore);
                }
                
            }
            //if algorymn = 2 then search for pairs etc
            if((Algorithm & 2) != 0)
                if (intLen1 > intLen2)
                {
                    FuzzyAlg2(String1, String2, ref intScore, ref intTotScore);
                }
                else
                {
                    FuzzyAlg2(String2, String1, ref intScore, ref intTotScore);
                }
            return intScore / intTotScore;
        }

        private void FuzzyAlg2(string string1, string string2, ref int intScore, ref int intTotScore)
        {
            int intLen1 = string1.Length;
            for (int intCurLen = 2; intCurLen < intLen1; intCurLen++)
            {
                string strWork = string2;
                int intTo = intLen1 - intCurLen + 1;
                intTotScore = intTotScore + (int)(intLen1 / intCurLen);
                for(int i = 1; i <= intTo; i += intCurLen)
                {
                    int intPos = strWork.IndexOf(string1.Substring(i-1, intCurLen));
                    if (intPos > 0)
                    {
                        //strWork.Substring(i, intPos) = intCurLen.ToString;
                        strWork.Remove(i-1, intPos).Insert(i-1, intCurLen.ToString());
                        intScore++;
                    }
                }
            }
        }

        private void FuzzyAlg1(string string1, string string2,ref int intScore, ref int intTotScore)
                {
                    int intLen1 = string1.Length;
                    intTotScore = intTotScore + intLen1;
                    int intPos = 0;
                    for (int i = 1; i < intLen1; i++)
                    {
                        int intStartPos = intPos + 1;
                        intPos = string2.IndexOf(string1.Substring(i-1, 1), intStartPos-1);
                        if (intPos > 0)
                        {
                            if (intPos > intStartPos + 3)
                            {
                                intPos = intStartPos;
                            }
                            else
                            {
                                intScore++;
                            }
                        }
                        else
                        {
                            intPos = intStartPos-1;
                        }
                    }
                }

            }
        }
     
